### PR TITLE
[api][SIMS-#227] Create ORM migration to setup institution auth table structure

### PR DESCRIPTION
**Pull request by fwpushan**
-----
1. Delete old pivot table institutions_users
2. Add new entity table institution_user
3. Add institution_user_auth table and entity model
3. Add code table for institution user types and roles
4. Create e2e tests to check all connection and relationship loading
----

N.B. This database setup for US #227

### CODE REVIEW

File: sources/packages/api/src-sql/AuthCodeTables/Create-initial-institution-user-types.sql [hunk 1]
```diff
@@ -0,0 +1,23 @@
+-- Inserting initial values for 'institution_user_type_roles' table. This will be part of table creation migration. So, no drop migration sql, this data will destroy with table in drop migration
```
Comment (andrewsignori-aot) [inline sources/packages/api/src-sql/AuthCodeTables/Create-initial-institution-user-types.sql:1]: Great work creating the descriptions, but it also shows that we need to duplicate a lot of the information if we are going to expand the table, as shown below. I would recommend revisiting the idea to have more two tables for users and roles types.

image

File: sources/packages/api/src-sql/AuthCodeTables/Create-institution-user-type.sql [hunk 2]
```diff
@@ -0,0 +1,40 @@
+CREATE TABLE IF NOT EXISTS institution_user_type_roles(
+  id SERIAL PRIMARY KEY,
+  user_type_description VARCHAR(500) NULL,
+  user_role_description VARCHAR(500) NULL,
```
Comment (guru-aot) [inline sources/packages/api/src-sql/AuthCodeTables/Create-institution-user-type.sql:4]: Can we please increase the size of the user_role_description to more like 1000, I guess postgres supports more than 8000 limit too.

File: sources/packages/api/src-sql/AuthCodeTables/Create-initial-institution-user-types.sql (hunk 1 above)
Comment (fwpushan) [reply sources/packages/api/src-sql/AuthCodeTables/Create-initial-institution-user-types.sql:1]: I agree ....

File: sources/packages/api/src-sql/AuthCodeTables/Create-institution-user-type.sql (hunk 2 above)
Comment (fwpushan) [reply sources/packages/api/src-sql/AuthCodeTables/Create-institution-user-type.sql:4]: yes, we can do that the actual limit is around 2^n - 1, where n is size of unicode char buffer.

Comment (fwpushan) [reply sources/packages/api/src-sql/AuthCodeTables/Create-institution-user-type.sql:4]: lets discuss and fix some good number for these kind of fields

File: sources/packages/api/src-sql/AuthCodeTables/Drop-institution-user-type.sql [hunk 3]
```diff
@@ -0,0 +1,2 @@
+-- DROP Table
+DROP TABLE IF EXISTS "institution_user_type_role";
```
Comment (andrewsignori-aot) [inline sources/packages/api/src-sql/AuthCodeTables/Drop-institution-user-type.sql:2]: I believe that there is an "s" missing in the name. The correct name of the table to be dropped is "institution_user_type_roles".

File: sources/packages/api/src-sql/InstitutionAuth/Create-institution-auth.sql [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  -- Audit columns
+  created_at timestamp without time zone NOT NULL DEFAULT now(),
+  updated_at timestamp without time zone NOT NULL DEFAULT now(),
+  creator INT NULL DEFAULT NULL REFERENCES users(id) ON DELETE SET NULL,
+  modifier INT NULL DEFAULT NULL REFERENCES users(id) ON DELETE SET NULL
+);
+
+COMMENT ON TABLE institution_user_auth IS 'Relation table to combine institution user and location with specific user type and roles';
+
+-- COMMENT ON COLUMN institution_user_auth.id IS '';
```
Comment (andrewsignori-aot) [inline sources/packages/api/src-sql/InstitutionAuth/Create-institution-auth.sql:28]: Please remove the commented line.

File: sources/packages/api/src/database/entities/institution.model.ts [hunk 5]
```diff
@@ -78,11 +72,5 @@ export class Institution extends RecordDataModel {
   })
   institutionAddress: InstitutionAddress;
 
-  @ManyToMany(() => User, { eager: false, cascade: true })
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/database/entities/institution.model.ts:81]: Can you elaborate a bit on why this relationship was removed?

File: sources/packages/api/src/database/__tests__/institution-user.e2e-spec.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  userFactory,
+  institutionUserTypeAndRoleFactory,
+  institutionFactory,
+  institutionUserFactory,
+  institutionLocationFactory,
+  institutionUserAuthFactory,
+} from "../factories";
+import { InstitutionLocation } from "../entities/institution-location.model";
+
+describe("Test student model", () => {
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/database/__tests__/institution-user.e2e-spec.ts:23]: Are these tests enable on CI/CD?

File: sources/packages/api/src/database/entities/institution-user-auth.model.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+import { RecordDataModel } from "./record.model";
+
+@Entity({
+  name: "institution_user_auth",
+})
+export class InstitutionUserAuth extends RecordDataModel {
+  @PrimaryGeneratedColumn()
+  id: number;
+
+  @ManyToOne((type) => InstitutionLocation, { eager: true })
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/database/entities/institution-user-auth.model.ts:14]: I would recommend removing the eager loading and letting the developer decide when it makes sense to retrieve the relational data.

File: sources/packages/api/src/database/entities/institution-user-auth.model.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  @JoinColumn({
+    name: "institution_location_id",
+    referencedColumnName: "id",
+  })
+  location: InstitutionLocation;
+
+  @ManyToOne((type) => InstitutionUserTypeAndRole, { eager: true })
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/database/entities/institution-user-auth.model.ts:21]: I would recommend removing the eager loading and letting the developer decide when it makes sense to retrieve the relational data and which columns are needed.

File: sources/packages/api/src/database/entities/institution-user.model.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+import { User } from "./user.model";
+
+@Entity({
+  name: "institution_users",
+})
+export class InstitutionUser extends RecordDataModel {
+  @PrimaryGeneratedColumn()
+  id: number;
+
+  @OneToOne((type) => User, { eager: true, cascade: true })
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/database/entities/institution-user.model.ts:20]: I would recommend removing the eager loading and letting the developer decide when it makes sense to retrieve the relational data.

File: sources/packages/api/src/database/entities/institution-user.model.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+  @JoinColumn({
+    name: "user_id",
+    referencedColumnName: "id",
+  })
+  user: User;
+
+  @OneToOne((type) => Institution, { eager: true, cascade: true })
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/database/entities/institution-user.model.ts:27]: I would recommend removing the eager loading and letting the developer decide when it makes sense to retrieve the relational data.

**Review by andrewsignori-aot — CHANGES_REQUESTED** [review]
Great work creating all the models and DB structures.
Please take a look in the comments.

File: sources/packages/api/src/database/entities/institution.model.ts (hunk 5 above)
Comment (fwpushan) [reply sources/packages/api/src/database/entities/institution.model.ts:81]: Currently-
1. There is Institution user entity  
2. Which connects institution, user, + location (optional) + auth type_role

File: sources/packages/api/src/database/__tests__/institution-user.e2e-spec.ts (hunk 6 above)
Comment (fwpushan) [reply sources/packages/api/src/database/__tests__/institution-user.e2e-spec.ts:23]: Removed from CI/CD

File: sources/packages/api/src/database/entities/institution-user.model.ts (hunk 9 above)
Comment (fwpushan) [reply sources/packages/api/src/database/entities/institution-user.model.ts:20]: Institution user is basically anchor entity connecting dots. So, I believe eager loading will be required here.

File: sources/packages/api/src/database/entities/institution-user-auth.model.ts (hunk 8 above)
Comment (fwpushan) [reply sources/packages/api/src/database/entities/institution-user-auth.model.ts:21]: I think eager loading for location is not required but TypeAndRole should be eager .... any development perspective I don't see any case where we need that entity without that type.

File: sources/packages/api/src/database/entities/institution-user.model.ts (hunk 10 above)
Comment (fwpushan) [reply sources/packages/api/src/database/entities/institution-user.model.ts:27]: Institution user is basically anchor entity connecting dots. So, I believe eager loading will be required here.

File: sources/packages/api/src-sql/AuthCodeTables/Create-initial-institution-user-types.sql (hunk 1 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src-sql/AuthCodeTables/Create-initial-institution-user-types.sql:1]: Team decision to go ahead and keep one table removing the descriptions.

File: sources/packages/api/src/database/entities/institution-user-type-role.model.ts [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+})
+export class InstitutionUserTypeAndRole extends RecordDataModel {
+  @PrimaryGeneratedColumn()
+  id: number;
+
+  @Column({
+    name: "user_type",
+    nullable: false,
+  })
+  type: string;
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/database/entities/institution-user-type-role.model.ts:15]: Can we create the string enums with the defined types?

File: sources/packages/api/src/database/entities/institution-user-type-role.model.ts [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+  @Column({
+    name: "user_role",
+    nullable: true,
+  })
+  role: string;
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/database/entities/institution-user-type-role.model.ts:21]: Can we create the string enums with the defined roles?

**Review by andrewsignori-aot — APPROVED** [review]
👍

File: sources/packages/api/src/database/entities/institution-user-type-role.model.ts (hunk 12 above)
Comment (fwpushan) [reply sources/packages/api/src/database/entities/institution-user-type-role.model.ts:21]: The enums are created but not attached with type def, I will do that in upcoming PR

**Pull request merged**